src/town_cmd.cpp

```diff
@@ -2784,36 +2809,68 @@ static bool BuildTownHouse(Town *t, TileIndex tile)
 			if (callback_res != CALLBACK_FAILED && !Convert8bitBooleanCallback(hs->grf_prop.grffile, CBID_HOUSE_ALLOW_CONSTRUCTION, callback_res)) continue;
 		}
 
-		/* build the house */
-		t->cache.num_houses++;
-
 		/* Special houses that there can be only one of. */
 		t->flags |= oneof;
 
-		uint8_t construction_counter = 0;
-		uint8_t construction_stage = 0;
+		BuildTownHouse(t, tile, hs, house, random_bits);
 
-		if (_generating_world || _game_mode == GM_EDITOR) {
-			uint32_t construction_random = Random();
+		return true;
+	}
 
-			construction_stage = TOWN_HOUSE_COMPLETED;
-			if (Chance16(1, 7)) construction_stage = GB(construction_random, 0, 2);
+	return false;
+}
 
-			if (construction_stage == TOWN_HOUSE_COMPLETED) {
-				ChangePopulation(t, hs->population);
-			} else {
-				construction_counter = GB(construction_random, 2, 2);
-			}
-		}
+CommandCost CmdPlaceHouse(DoCommandFlag flags, TileIndex tile, HouseID house)
+{
+	if (_game_mode != GM_EDITOR) return CMD_ERROR;
+	if (Town::GetNumItems() == 0) return_cmd_error(STR_ERROR_MUST_FOUND_TOWN_FIRST);
 
-		MakeTownHouse(tile, t, construction_counter, construction_stage, house, random_bits);
-		UpdateTownRadius(t);
-		UpdateTownGrowthRate(t);
+	if (static_cast<size_t>(house) >= HouseSpec::Specs().size()) return CMD_ERROR;
+	const HouseSpec *hs = HouseSpec::Get(house);
+	if (!hs->enabled) return CMD_ERROR;
 
-		return true;
+	if (TimerGameCalendar::year < hs->min_year || TimerGameCalendar::year > hs->max_year) return CMD_ERROR;
```

Comment:
They are very similar but with a few differences.

Some of the tests aren't performed because we let the user do more than towns.
TryBuildTownHouse just continues its loop, by CmdPlaceHouse needs to return error strings.

I did consider making TryBuildTownHouse issue a Command<CMD_PLACE_HOUSE>::Do. but that complicates CmdPlaceHouse more -- e.g TryBuildTownHouse only evaluates the slope and some other things once, not every attempt to build.